### ISSUE TITLE
BART metrics.json and validation checkpoint metrics seem to disagree

### ISSUE THREAD

**Issue by vikigenius** [huggingface/transformers#7303]
## Environment info

     
- `transformers` version: 3.1.0
- Platform: Linux 4.14
- Python version:3.7.9
- PyTorch version (GPU?):1.6.0 (True)
- Using GPU in script?: True
- Using distributed or parallel set-up in script?: True

### Who can help

 examples/seq2seq: @sshleifer 
## Information

Model I am using (BART):

The problem arises when using:
* [*] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [*] my own task or dataset: (give details below)
I am using a small subset of https://github.com/pubmedqa/pubmedqa to generate questions from the given passage.

## To reproduce

1. Do distributed training
2. Run the finetune script with appropriate arguments.

The saved checkpoint says:
val_avg_rouge2=13.0975-step_count=4.ckpt

However the metrics.json file says that val_avg_rouge2=6.46015 and there are better step_counts in comparison.

## Expected behavior

The best metric in metrics.json should be chosen for saving checkpoints.

**Comment by sshleifer** [comment]
I can replicate in pl 0.8.5, great catch.

Smaller command to replicate:

```
export MAX_LEN=128
export m=export m=sshleifer/student_marian_en_ro_6_3
python finetune.py \
  --learning_rate=3e-4 \
  --do_train \
  --do_predict \
  --fp16 \
  --val_check_interval 0.25 \
  --data_dir $ENRO_DIR \
  --max_source_length $MAX_LEN --max_target_length $MAX_LEN --val_max_target_length $MAX_LEN --test_max_target_length $MAX_LEN \
  --freeze_encoder --freeze_embeds \
  --train_batch_size=64 --eval_batch_size=64 \
  --tokenizer_name $m --model_name_or_path $m \
  --warmup_steps 500 --sortish_sampler --logger_name wandb \
  --fp16_opt_level=O1 --task translation --num_sanity_val_steps=0 \
  --model_name_or_path $m --gpus 8 --num_train_epochs=12 \
  --data_dir wmt_en_ro --output_dir dmar_pl_only_v2 --save_top_k=10
```

What I think is happening is that only information from 1 gpu is being used either to save `metrics.json` or to save checkpoints. Hopefully the former.

Trying now in later version of PL.

**Comment by sshleifer** [comment]
Every single rank is saving checkpoints.

**Comment by vikigenius** [comment]
@sshleifer Wow, so is that like a race condition where the last RANK "wins"? What about the weights of model? Would they be same across all the ranks, or would they be a problem as well?

**Comment by sshleifer** [comment]
You should discuss here:
https://github.com/PyTorchLightning/pytorch-lightning/issues/3597

This is my best understanding of the issue.

**Comment by sshleifer** [comment]
Weights will be the same across all ranks. You just have suboptimal checkpoint saving logic. You can kind of workaround by passing `--save_top_k=5` and then manually picking which one you like by looking at metrics.json.
